# 15 · GSE232381 · bulk_RNA_seq · are the GSE65391 modules present here?

Reads this study's `expression.rds` and the GSE65391 end product `module_genes.csv` (gene names and
module labels only; no GSE65391 expression values). Writes `data/run_artifacts/GSE232381/array_module_preservation.csv`.

Module preservation (Langfelder P et al. *PLoS Comput Biol* 2011;7:e1001057) asks whether genes that
form a module in one study also hang together in another. The full method also needs the reference
study's expression data. Here only the gene lists travel, so we use the two **density** statistics,
which need only this study's data:

| statistic | meaning |
|---|---|
| mean_cor | mean Pearson correlation over all pairs of the module's genes, in this study |
| prop_var_PC1 | share of the module genes' variance explained by their first principal component |

**Test.** For each array module, the same statistics for 1,000 random gene sets of the same size drawn
from this study's genes. Z = (observed - mean of random) / SD of random. **Reading**, borrowed from
Langfelder 2011 for Zsummary: Z > 10 strong evidence, 2-10 weak to moderate, < 2 none.

## Standardize this study's genes

In [1]:
source("../src/paths.R")
x  <- readRDS(art("GSE232381", "expression.rds"))
mg <- read.csv(art("GSE65391", "module_genes.csv"))
Zs <- scale(t(x$E))                         # samples x genes, each gene standardised in this study
Zs <- Zs[, colSums(is.na(Zs)) == 0]
c(samples = nrow(Zs), genes = ncol(Zs))

samples   genes 
     16   16605

**Explanation**
- `module_genes.csv` of GSE65391 (notebook 05) lists the array modules: gene and module only, no expression
  values.
- `scale(t(x$E))` standardizes each gene over this study's 16 samples: samples by genes.
- `colSums(is.na(Zs)) == 0` drops genes that do not vary, whose standardized values are missing.

## Preservation statistics with a random-gene null

In [2]:
stat <- function(Z) {
  p <- ncol(Z)
  c(mean_cor = (var(rowSums(Z)) - p) / (p * (p - 1)),        # sum of all pairwise correlations of standardised genes
    prop_var_PC1 = svd(Z, nu = 0, nv = 0)$d[1]^2 / sum(svd(Z, nu = 0, nv = 0)$d^2))
}
set.seed(SEED)
pres <- do.call(rbind, lapply(split(mg$gene, mg$module), function(g) {
  g_here <- intersect(g, colnames(Zs))
  obs  <- stat(Zs[, g_here, drop = FALSE])
  null <- replicate(1000, stat(Zs[, sample(colnames(Zs), length(g_here)), drop = FALSE]))
  z    <- (obs - rowMeans(null)) / apply(null, 1, sd)
  data.frame(genes_in_array_module = length(g), genes_measured_here = length(g_here),
             mean_cor = obs[1], Z_mean_cor = z[1], prop_var_PC1 = obs[2], Z_prop_var_PC1 = z[2])
}))
pres$module <- rownames(pres)
pres$reading <- cut(pmin(pres$Z_mean_cor, pres$Z_prop_var_PC1), c(-Inf, 2, 10, Inf),
                    labels = c("none", "weak to moderate", "strong"))
pres <- pres[order(-pres$Z_mean_cor), c("module", "genes_in_array_module", "genes_measured_here", "mean_cor",
                                          "Z_mean_cor", "prop_var_PC1", "Z_prop_var_PC1", "reading")]
format(pres, digits = 3)

,module,genes_in_array_module,genes_measured_here,mean_cor,Z_mean_cor,prop_var_PC1,Z_prop_var_PC1,reading
,<I<chr>>,<I<chr>>,<I<chr>>,<I<chr>>,<I<chr>>,<I<chr>>,<I<chr>>,<I<chr>>
blue,blue,1557,1415,0.529,404.6,0.618,39.18,strong
turquoise,turquoise,3126,2922,0.197,233.2,0.330,-1.73,none
brown,brown,647,564,0.386,169.3,0.565,18.83,strong
purple,purple,152,149,0.624,112.6,0.681,15.85,strong
greenyellow,greenyellow,141,134,0.679,110.9,0.728,15.96,strong
red,red,295,281,0.332,92.3,0.478,8.33,weak to moderate
black,black,280,263,0.316,79.4,0.418,4.52,weak to moderate
pink,pink,212,187,0.373,77.8,0.445,5.16,weak to moderate
yellow,yellow,383,278,0.271,77.7,0.401,3.81,weak to moderate


**Explanation**
- `stat(Z)` returns two statistics for a set of standardized genes:
  - `mean_cor`: the mean correlation over all pairs of genes. It is computed from the variance of the row
    sums: for standardized genes, the variance of their sum is p plus the sum of all pairwise correlations.
  - `prop_var_PC1`: the share of the variance explained by the first principal component, from the singular
    values `svd(Z)$d`.
- For each array module (`split(mg$gene, mg$module)`):
  - `g_here` keeps the module genes measured in this study;
  - `obs` gives the two statistics for them;
  - `null` gives the same two statistics for 1,000 random sets of the same number of genes;
    `replicate(1000, ...)` repeats the random draw;
  - `z` is how many standard deviations the observed value lies above the random ones.
- `reading` uses the smaller of the two Z values: below 2 none, 2 to 10 weak to moderate, above 10 strong.
  These are the thresholds of the WGCNA preservation Z summary (Langfelder P, Luo R, Oldham MC, Horvath S.
  Is my network module preserved and reproducible? *PLoS Comput Biol* 2011;7:e1001057).

**Result.** By both statistics (the reading uses the smaller Z):
- **strong:** blue, brown, purple, greenyellow;
- **weak to moderate:** red, black (interferon), pink (plasma cell), yellow, magenta, green, salmon
  (monocyte), tan, cyan (neutrophil granule);
- **none:** turquoise.

Z for mean_cor grows with module size, because the null for a large random gene set varies little.
That is why turquoise (2,922 genes measured) has Z_mean_cor = 233 but a negative Z for prop_var_PC1.

## Save

In [3]:
write.csv(pres, art("GSE232381", "array_module_preservation.csv"), row.names = FALSE)

**Explanation**
- `array_module_preservation.csv` keeps the table for the federation notebooks.